Userland/Libraries/LibVideo/DecoderError.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2022, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/Format.h>
+#include <AK/String.h>
+#include <errno.h>
+
+namespace Video {
+
+struct DecoderError;
+
+template<typename T>
+using DecoderErrorOr = ErrorOr<T, DecoderError>;
+
+enum class DecoderErrorCategory : u32 {
+    Unknown = 0,
+    IO,
+    // The input is corrupted.
+    Corrupted,
+    // The input uses features that are not yet implemented.
+    NotImplemented,
+};
+
+struct DecoderError {
+public:
+    static DecoderError with_description(DecoderErrorCategory category, StringView description)
+    {
+        return DecoderError(category, description);
+    }
+
+    template<typename... Parameters>
+    static DecoderError format(DecoderErrorCategory category, CheckedFormatString<Parameters...>&& format_string, Parameters const&... parameters)
+    {
+        AK::VariadicFormatParams variadic_format_params { parameters... };
+        return DecoderError::with_description(category, String::vformatted(format_string.view(), variadic_format_params));
+    }
+
+    DecoderErrorCategory category()
+    {
+        return m_category;
+    }
+
+    StringView description()
+    {
+        return m_description;
+    }
+
+    StringView string_literal()
+    {
+        return m_description;
+    }
+
+    DecoderError with_caller(String caller)
+    {
+        return DecoderError::format(m_category, "{}: {}", caller, m_description);
+    }
+
+private:
+    DecoderError(DecoderErrorCategory category, String description)
+        : m_category(category)
+        , m_description(move(description))
+    {
+    }
+
+    DecoderErrorCategory m_category { DecoderErrorCategory::Unknown };
+    String m_description { String::empty() };
```

Comment:
Isn't using the default constructor here better?
```suggestion
    String m_description;
```

Reply:
Fixed.